### PR TITLE
EID-1877 Decouple Stub Connector from the HSM

### PR DESCRIPTION
**Pull request by idrop**
_TLDR; The Stub Connector represents a stub country. It should not have its saml signing coupled to an HSM key._

---

See https://govukverify.atlassian.net/browse/EID-1877

The Stub Connector (a stub country) currently has its metadata certs+keys generated via the verify-metadata-controller, and the saml signing key is held in the HSM, see https://test-integration-connector.london.verify.govsvc.uk/ConnectorMetadata

Every SAML request from stub connector is currently signed by a key in the CloudHSM. This was done as a convenience as the truststore for this signing is also generated via verify-metadata-controller, and easily made available to the ESP (the service that reads the connector metadata).

We want to model a a stub country that has its own PKI completely outside the HSM. The eventual goal is to run the Stub Connector in PaaS, to complete the separation.

Also, the Stub Connector is holding on to HSM key handles when it generates SAML requests.
This contributes to our quota of a max of ~ 3300 handles.

This PR will:

* Create a pipeline to set a metadata signing cert+key, saml signing cert+key plus a metadata signing truststore, all from verify-dev-pki, into a cluster-wide configmap named test-pki-configmap
* Read test-pki-configmap to set a truststore in ESP
* Read test-pki-configmap via a volumeMount to provide metadata and saml signing credentials to the connector
* Remove HSM client config from stub connector
* Publish connector node metadata from a Resource using a mustache template. The method is equivalent to how connector node metadata is currently created in the verify-metadata-controller
* Use sensible default values in connector config.yml to avoid setting config via env vars

You can see connector node metadata created with a cherry pick of this commit at https://stub-connector.eidas.test.london.sandbox.govsvc.uk/ConnectorMetadata

Out of Scope: 
* Create two instances of stub country, on the master branch
* Remove current process for publishing metadata within verify-metadata-controller
* Publish the metadata signing CA certs, see https://test-integration-connector.london.verify.govsvc.uk/ConnectorMetadataSigningCertificates
* Migrate to PaaS

### CODE REVIEW

File: stub-connector/src/main/java/uk/gov/ida/notification/stubconnector/resources/ConnectorNodeCredentialConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
+package uk.gov.ida.notification.stubconnector.resources;
```
Comment (Wynndow) [inline stub-connector/src/main/java/uk/gov/ida/notification/stubconnector/resources/ConnectorNodeCredentialConfiguration.java]: Is this the correct place for this? Perhaps somewhere like `uk.gov.ida.notification.stubconnector.configuration`

File: stub-connector/src/main/java/uk/gov/ida/notification/stubconnector/resources/MetadataResource.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+    private final MetadataGenerator metadataGenerator;
+
+    public MetadataResource(MetadataGenerator metadataGenerator) {
+        this.metadataGenerator = metadataGenerator;
+    }
+
+    @GET
+    public Response getConnectorNodeMetadata() throws Exception {
+        XMLObject xmlObject = metadataGenerator.getConnectorMetadata();
```
Comment (Wynndow) [inline stub-connector/src/main/java/uk/gov/ida/notification/stubconnector/resources/MetadataResource.java]: Rather than regenerate and sign the metadata every time this endpoint is hit, could it be cached? Or just generate it once when the `MetadataResource` is initialised, with a really long validity period?

File: chart/templates/esp-deployment.yaml [hunk 3]
```diff
@@ -70,14 +70,11 @@ spec:
 {{- if .Values.stubConnector.enabled }}
         - name: CONNECTOR_NODE_METADATA_TRUSTSTORE
           valueFrom:
-            secretKeyRef:
-              name: {{ .Release.Name }}-connector-metadata
-              key: metadataSigningTruststoreBase64
+            configMapKeyRef:
```
[the remark this answers is not in the document]
Comment (idrop) [reply chart/templates/esp-deployment.yaml]: Mainly so that we can see exactly how the truststore, plus keys and certs were created.
The fact that the configmap resolves to files in the `/pki` directory when volume mounted makes it even more attractive.
Granted this would be obsolete if/when we move to PaaS.

File: stub-connector/src/main/java/uk/gov/ida/notification/stubconnector/resources/ConnectorNodeCredentialConfiguration.java (hunk 1 above)
Comment (idrop) [reply stub-connector/src/main/java/uk/gov/ida/notification/stubconnector/resources/ConnectorNodeCredentialConfiguration.java]: Agree! Moving

File: stub-connector/src/main/java/uk/gov/ida/notification/stubconnector/resources/MetadataResource.java (hunk 2 above)
Comment (idrop) [reply stub-connector/src/main/java/uk/gov/ida/notification/stubconnector/resources/MetadataResource.java]: As this is the stub connector I wasn't too concerned with caching, but I'm going to add a `@CacheControl` annotation to simulate a cache. Also adding a metadata expiry env var so we can simulate a stub country with expired metadata.

Comment (Wynndow) [reply stub-connector/src/main/java/uk/gov/ida/notification/stubconnector/resources/MetadataResource.java]: I guess I was thinking more that the metadata isn't going to change or be regenerated once the app is up and running, so to save processing does it make sense to just generate the metadata once when the app boots up, rather than doing it every time the endpoint is hit. The cache you've added will help but it might be better to just only do it once but with a long validity.

Having said this, the amount it'll get used means this probably doesn't matter at all.

File: proxy-node-shared/src/main/java/uk/gov/ida/notification/saml/ResponseAssertionDecrypter.java [hunk 4]
```diff
@@ -18,10 +16,6 @@
 
     public ResponseAssertionDecrypter(Credential credential) {
         this.decrypter = new DecrypterFactory().createDecrypter(Collections.singletonList(credential));
-        if (credential.getEntityId() != null && credential.getEntityId().equals(CloudHsmCredentialConfiguration.ID)) {
-            ProxyNodeLogger.info("Using CloudHSM so set JCA provider to Cavium");
-            this.decrypter.setJCAProviderName("Cavium");
```
Comment (galund) [inline proxy-node-shared/src/main/java/uk/gov/ida/notification/saml/ResponseAssertionDecrypter.java]: Are we sure we no longer need this? The translator also uses the `ResponseAssertionDecrypter` (maybe)?

File: ci/build/test-pki.yaml [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+      type: git
+      source:
+        uri: https://github.com/alphagov/verify-dev-pki.git
+
+    jobs:
+
+    - name: build-test-pki
+      plan:
+      - get: verify-dev-pki
+        trigger: true
```
[the remark this answers is not in the document]
Comment (idrop) [reply ci/build/test-pki.yaml]: Removed the trigger, we can invoke by hand

File: ci/sandbox/test-pki.yaml [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+      type: git
+      source:
+        uri: https://github.com/alphagov/verify-dev-pki.git
+
+    jobs:
+
+    - name: build-test-pki
+      plan:
+      - get: verify-dev-pki
+        trigger: true
```
[the remark this answers is not in the document]
Comment (idrop) [reply ci/sandbox/test-pki.yaml]: Removed the trigger, we can invoke by hand

File: proxy-node-shared/src/main/java/uk/gov/ida/notification/saml/ResponseAssertionDecrypter.java (hunk 4 above)
Comment (galund) [reply proxy-node-shared/src/main/java/uk/gov/ida/notification/saml/ResponseAssertionDecrypter.java]: (Nah, that's only in a test, and has a non-HSM key as well.)

**Review by galund — APPROVED** [review]

**Review by Wynndow — APPROVED** [review]
